{
  "origin": "codeshovel",
  "repositoryName": "hadoop",
  "repositoryPath": "/home/shaiful/research/codeshovel/codeshovel-projects/hadoop/.git",
  "startCommitName": "HEAD",
  "sourceFileName": "ExceptionHandler.java",
  "functionName": "toCause",
  "functionId": "toCause___e-Exception",
  "sourceFilePath": "hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/web/resources/ExceptionHandler.java",
  "functionStartLine": 48,
  "functionEndLine": 67,
  "numCommitsSeen": 7,
  "timeTaken": 1538,
  "changeHistory": [
    "55a1defa503120bdceae52d45331e263a3934a63",
    "09a156fcce2bc1be4081717bf7ef7d290e80d818"
  ],
  "changeHistoryShort": {
    "55a1defa503120bdceae52d45331e263a3934a63": "Ybodychange",
    "09a156fcce2bc1be4081717bf7ef7d290e80d818": "Yintroduced"
  },
  "changeHistoryDetails": {
    "55a1defa503120bdceae52d45331e263a3934a63": {
      "type": "Ybodychange",
      "commitMessage": "HDFS-6475. WebHdfs clients fail without retry because incorrect handling of StandbyException. Contributed by Yongjun Zhang.\n\n\ngit-svn-id: https://svn.apache.org/repos/asf/hadoop/common/trunk@1605217 13f79535-47bb-0310-9956-ffa450edef68\n",
      "commitDate": "24/06/14 4:07 PM",
      "commitName": "55a1defa503120bdceae52d45331e263a3934a63",
      "commitAuthor": "Aaron Myers",
      "commitDateOld": "31/03/14 12:25 AM",
      "commitNameOld": "004d0854b7964d4f748f6e91b2d54a84928843f7",
      "commitAuthorOld": "Arun Murthy",
      "daysBetweenCommits": 85.65,
      "commitsBetweenForRepo": 513,
      "commitsBetweenForFile": 1,
      "diff": "@@ -1,7 +1,20 @@\n   private static Exception toCause(Exception e) {\n-    final Throwable t \u003d e.getCause();\n-    if (t !\u003d null \u0026\u0026 t instanceof Exception) {\n-      e \u003d (Exception)e.getCause();\n+    final Throwable t \u003d e.getCause();    \n+    if (e instanceof SecurityException) {\n+      // For the issue reported in HDFS-6475, if SecurityException\u0027s cause\n+      // is InvalidToken, and the InvalidToken\u0027s cause is StandbyException,\n+      // return StandbyException; Otherwise, leave the exception as is,\n+      // since they are handled elsewhere. See HDFS-6588.\n+      if (t !\u003d null \u0026\u0026 t instanceof InvalidToken) {\n+        final Throwable t1 \u003d t.getCause();\n+        if (t1 !\u003d null \u0026\u0026 t1 instanceof StandbyException) {\n+          e \u003d (StandbyException)t1;\n+        }\n+      }\n+    } else {\n+      if (t !\u003d null \u0026\u0026 t instanceof Exception) {\n+        e \u003d (Exception)t;\n+      }\n     }\n     return e;\n   }\n\\ No newline at end of file\n",
      "actualSource": "  private static Exception toCause(Exception e) {\n    final Throwable t \u003d e.getCause();    \n    if (e instanceof SecurityException) {\n      // For the issue reported in HDFS-6475, if SecurityException\u0027s cause\n      // is InvalidToken, and the InvalidToken\u0027s cause is StandbyException,\n      // return StandbyException; Otherwise, leave the exception as is,\n      // since they are handled elsewhere. See HDFS-6588.\n      if (t !\u003d null \u0026\u0026 t instanceof InvalidToken) {\n        final Throwable t1 \u003d t.getCause();\n        if (t1 !\u003d null \u0026\u0026 t1 instanceof StandbyException) {\n          e \u003d (StandbyException)t1;\n        }\n      }\n    } else {\n      if (t !\u003d null \u0026\u0026 t instanceof Exception) {\n        e \u003d (Exception)t;\n      }\n    }\n    return e;\n  }",
      "path": "hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/web/resources/ExceptionHandler.java",
      "extendedDetails": {}
    },
    "09a156fcce2bc1be4081717bf7ef7d290e80d818": {
      "type": "Yintroduced",
      "commitMessage": "HDFS-2539. Support doAs and GETHOMEDIRECTORY in webhdfs.\n\n\ngit-svn-id: https://svn.apache.org/repos/asf/hadoop/common/trunk@1200731 13f79535-47bb-0310-9956-ffa450edef68\n",
      "commitDate": "10/11/11 8:19 PM",
      "commitName": "09a156fcce2bc1be4081717bf7ef7d290e80d818",
      "commitAuthor": "Tsz-wo Sze",
      "diff": "@@ -0,0 +1,7 @@\n+  private static Exception toCause(Exception e) {\n+    final Throwable t \u003d e.getCause();\n+    if (t !\u003d null \u0026\u0026 t instanceof Exception) {\n+      e \u003d (Exception)e.getCause();\n+    }\n+    return e;\n+  }\n\\ No newline at end of file\n",
      "actualSource": "  private static Exception toCause(Exception e) {\n    final Throwable t \u003d e.getCause();\n    if (t !\u003d null \u0026\u0026 t instanceof Exception) {\n      e \u003d (Exception)e.getCause();\n    }\n    return e;\n  }",
      "path": "hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/web/resources/ExceptionHandler.java"
    }
  }
}